source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h

```diff
@@ -15,21 +15,53 @@ namespace Extensions {
 namespace UdpFilters {
 namespace UdpProxy {
 
+/**
+ * All UDP proxy stats. @see stats_macros.h
+ */
+#define ALL_UDP_PROXY_STATS(COUNTER, GAUGE)                                                        \
+  COUNTER(downstream_sess_rx_bytes)                                                                \
+  COUNTER(downstream_sess_rx_datagrams)                                                            \
+  COUNTER(downstream_sess_rx_errors)                                                               \
+  COUNTER(downstream_sess_total)                                                                   \
+  COUNTER(downstream_sess_tx_bytes)                                                                \
+  COUNTER(downstream_sess_tx_datagrams)                                                            \
+  COUNTER(downstream_sess_tx_errors)                                                               \
+  COUNTER(idle_timeout)                                                                            \
+  GAUGE(downstream_sess_active, Accumulate)
+
+/**
+ * Struct definition for all UDP proxy stats. @see stats_macros.h
+ */
+struct UdpProxyStats {
+  ALL_UDP_PROXY_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
 class UdpProxyFilterConfig {
 public:
   UdpProxyFilterConfig(Upstream::ClusterManager& cluster_manager, TimeSource& time_source,
+                       Stats::Scope& root_scope,
                        const envoy::config::filter::udp::udp_proxy::v2alpha::UdpProxyConfig& config)
-      : cluster_manager_(cluster_manager), time_source_(time_source), config_(config) {}
+      : cluster_manager_(cluster_manager), time_source_(time_source), cluster_(config.cluster()),
+        session_timeout_(PROTOBUF_GET_MS_OR_DEFAULT(config, idle_timeout, 60 * 1000)),
+        stats_(generateStats(config.stat_prefix(), root_scope)) {}
 
-  Upstream::ThreadLocalCluster* getCluster() const {
-    return cluster_manager_.get(config_.cluster());
-  }
+  Upstream::ThreadLocalCluster* getCluster() const { return cluster_manager_.get(cluster_); }
+  std::chrono::milliseconds sessionTimeout() const { return session_timeout_; }
+  UdpProxyStats& stats() const { return stats_; }
   TimeSource& timeSource() const { return time_source_; }
 
 private:
+  static UdpProxyStats generateStats(const std::string& stat_prefix, Stats::Scope& scope) {
+    const auto final_prefix = fmt::format("udp.{}", stat_prefix);
+    return {ALL_UDP_PROXY_STATS(POOL_COUNTER_PREFIX(scope, final_prefix),
+                                POOL_GAUGE_PREFIX(scope, final_prefix))};
+  }
+
   Upstream::ClusterManager& cluster_manager_;
   TimeSource& time_source_;
-  const envoy::config::filter::udp::udp_proxy::v2alpha::UdpProxyConfig config_;
+  const std::string cluster_;
```

Comment:
Since we latch cluster_ here, can we remove cluster_manager_ from this class?

Reply:
We still need the cluster manager to fetch the thread local cluster info. I also need this for my follow up change.